### PR TITLE
Fix issue when using the multi line codec

### LINKED ISSUE #1: Make it work with multiline codec
Problem code: https://github.com/logstash-plugins/logstash-input-lumberjack/blob/bd3207e90d2c1121f5f4f814f6ec336facf49095/lib/logstash/inputs/lumberjack.rb#L47-L53

The lumberjack input cannot currently be used with the multiline codec due to the fact that the codec is shared across all lumberjack streams. This causes weird and confusing behavior that events from unrelated log streams are merged into a single event.

### PR DESCRIPTION
**Pull request by purbon**
Made the codec server and connection independant, so there is no issue when using for example the multiline codec.

Fixes #1 

Relates and depends to: https://github.com/elasticsearch/logstash-forwarder/pull/364

Probably not the best solution, but a good workaround for now.

### CODE REVIEW

**Comment by purbon** [comment]
@jordansissel need more work on the test side.

File: lib/logstash/inputs/lumberjack.rb [hunk 1]
```diff
@@ -33,7 +33,7 @@ class LogStash::Inputs::Lumberjack < LogStash::Inputs::Base
 
   public
   def register
-    require "lumberjack/server"
+    require "logstash/inputs/lumberjack/server"
```
Comment (suyograo) [inline lib/logstash/inputs/lumberjack.rb]: why this change?

Comment (purbon) [inline lib/logstash/inputs/lumberjack.rb]: Because the old lumberjack/server was not codec aware, basically with this change we need to create a new codec instance per connection.

This change can be done in the same ```lumberjack/server``` code, but after talking with @jordansissel we agreed that makes more sense to have this logic in here, so I had to redefine the server logic to accept a new codec in this input.

File: lib/logstash/inputs/lumberjack/server.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+    def data(sequence, map, &block)
+      block.call(map, @codec)
+      super(sequence, map)
+    end
+  end
+
+
+  class Server
+    def run(codec, &block)
```
Comment (jordansissel) [inline lib/logstash/inputs/lumberjack/server.rb]: I don't think we need to override Lumberjack::Server, I'll comment elsehere to show why

File: lib/logstash/inputs/lumberjack.rb [hunk 3]
```diff
@@ -43,10 +43,10 @@ def register
 
   public
   def run(output_queue)
-    @lumberjack.run do |l|
-      @codec.decode(l.delete("line")) do |event|
+    @lumberjack.run(@codec.clone) do |lines, codec|
```
Comment (jordansissel) [inline lib/logstash/inputs/lumberjack.rb]: this has a bug because @codec.clone is only invoked once, so each thread shares that clone.

We can avoid monkeypatching Lumberjack::Server (as mentioned in a comment below) and use `#accept` in a loop, dispatching threads ourselves:

```ruby
while true # or not shutting down?
  connection = @lumberjack.accept # blocks until accept
  Thread.new(connection, @codec.clone) do |connection, codec|
    connection.run do |line|
      codec.decode(line.delete("line")) do |event|
        ...
      end
    end
  end
end
```

(Code above should probably not just be in one method, but split out for easier testing.)

Comment (purbon) [inline lib/logstash/inputs/lumberjack.rb]: Sorry, fixing this right now.

**Comment by jordansissel** [comment]
new code lgtm; haven't tested yet.

**Comment by jordansissel** [comment]
specs pass, but I still need to review and test manually.

**Comment by jordansissel** [comment]
Tested manually, the plugin seems to terminate immediately upon startup:

```
Starting lumberjack input listener {:address=>"0.0.0.0:5043", :level=>:info, :file=>"logstash/inputs/lumberjack.rb", :line=>"38", :method=>"register"}
Pipeline started {:level=>:info, :file=>"logstash/pipeline.rb", :line=>"83", :method=>"run"}
Plugin is finished {:plugin=><LogStash::Inputs::Lumberjack ssl_certificate=>"../logstash-forwarder/server.crt", ssl_key=>"../logstash-forwarder/server.key", host=>"0.0.0.0">, :level=>:info, :file=>"logstash/plugin.rb", :line=>"61", :method=>"finished"}
```

I'm digging. Will provide a patch to fix it shortly.

**Comment by jordansissel** [comment]
```
Exception in lumberjack input {:exception=>#<NoMethodError: undefined method `accept' for #<Lumberjack::Server:0x2c69c4e9>>, :level=>:error, :file=>"logstash/inputs/lumberjack.rb", :line=>"58", :method=>"run"}
```

Ahha, I think I need to upgrade jls-lumberjack

**Comment by jordansissel** [comment]
```diff
diff --git a/lib/logstash/inputs/lumberjack.rb b/lib/logstash/inputs/lumberjack.rb
index fb4b80a..9b54ee2 100644
--- a/lib/logstash/inputs/lumberjack.rb
+++ b/lib/logstash/inputs/lumberjack.rb
@@ -45,16 +45,17 @@ class LogStash::Inputs::Lumberjack < LogStash::Inputs::Base
   def run(output_queue)
     while true do
       accept do |connection, codec|
-        invoke(connection, codec) do |_codec, lines|
-          _codec.decode(lines) do |event|
+        invoke(connection, codec) do |_codec, line, fields|
+          _codec.decode(line) do |event|
             decorate(event)
-            lines.each { |k,v| event[k] = v; v.force_encoding(Encoding::UTF_8) }
+            fields.each { |k,v| event[k] = v; v.force_encoding(Encoding::UTF_8) }
             output_queue << event
           end
         end
       end
     end
-  rescue
+  rescue => e
+    @logger.error("Exception in lumberjack input", :exception => e)
     shutdown(output_queue)
   end # def run

@@ -67,8 +68,8 @@ class LogStash::Inputs::Lumberjack < LogStash::Inputs::Base
   private
   def invoke(connection, codec, &block)
     Thread.new(connection, codec) do |_connection, _codec|
-      _connection.run do |lines|
-        block.call(_codec, lines.delete("line"))
+      _connection.run do |fields|
+        block.call(_codec, fields.delete("line"), fields)
       end
     end
   end
```

**Comment by jordansissel** [comment]
The above patch fixes a few things:
* It logs the error that occurred that would cause a shutdown
* it fixes a `NoMethodError: undefined method `each' for \"...\":String` error caused by `lines.each` which was a string prior to the above patch
* In the `block.call` of `#invoke` it sends the line and the fields separately now. Before this, the fields were being lost.

Otherwise, this patch looks great! I have tested manually with 2 Logstash forwarders and 1 logstash and using multiline codec and things work as expected.

I'd love a more comprehensive test suite to cover the problems I found, but am willing to wait until another day (and another PR) to have that land.

@purbon Can you apply the patch in the above comment and let update the PR? I think you can do this by doing running `patch -p1` from the logstash-input-lumberjack directory and pasting in the full diff text, then hitting ^D. :)

**Comment by jordansissel** [comment]
LGTM
